### ISSUE TITLE
Abstract Constructor Assignability Not Restricted Enough

### ISSUE THREAD

**Issue by aozgaa** [Microsoft/TypeScript#4190]
Consider the following code snippet:

```ts
abstract class A {}
class B extends A {}

var A2: new () => A;
var BCtor: typeof B;

A2 = A;
BCtor = A1;

```

The first assignment should fail because `A2` is new-able (it isn't an abstract constructor type), but no error is reported.

EDIT: simplified the sample.

**Comment by tinganho** [comment]
If this is a bug how do I do this https://github.com/Microsoft/TypeScript/issues/2947#issuecomment-120579600?

>@RyanCavanaugh I have another question about abstract classes. I wonder how I define a function that accepts a derived abstract class that is not abstract and the function should also "new" it.
 
>The following doesn't work:
 
>```typescript
function testDerived<T extends typeof ComposerComponent>(DerivedClass: T) {
   new DerivedClass(); // error
}
function testDerived(DerivedClass: typeof ComposerComponent) {
   new DerivedClass(); // error
}
```

And @RyanCavanaugh suggested:
>function testDerived(DerivedClass: new() => ComposerComponent) {


I think that there should be a modifier for declaring derivative of an abstract class that is not abstract.

```typescript
var A2: new () => implementationof A; // use m
```

**Comment by aozgaa** [comment]
Let me check my understanding. Suppose our class hierarchy looks like
```
    A
   /  \
 B    C 
```
where `A` is abstract but `B`,`C` are not.

Then you want to have an object `AA` which has a type that allows
```ts
AA = B;
AA = C;
new AA;
```
but forbids
```ts
AA = A;
```

If my understanding is correct, that's precisely the capability that should be provided by 
`var A2: new () => A;` in the example above. Right now, the forbidden condition isn't forbidden.

If I'm misunderstanding, can you clarify where I'm going astray?

**Comment by tinganho** [comment]
>If my understanding is correct, that's precisely the capability that should be provided by 
var A2: new () => A; in the example above. Right now, the forbidden condition isn't forbidden.

This is the only way that I know of that can do the capability we are talking about. 

Though personally I also feel that this is a bug. Though I want to have the capability to declare a non abstract class declaration type. If you fix this bug, then I don't know how to do it(except using `any`).

**Comment by aozgaa** [comment]
Could you clarify what you mean by "a non abstract class declaration type"?

**Comment by tinganho** [comment]
@aozgaa don't know the real name of it. but just to clarify my meaning:
```typescript
abstract class A {}
typeof A // abstract class declaration type
class B extends A {}
typeof B //non-abstract class declaration type
```

**Comment by aozgaa** [comment]
I see. So what's missing from the constructor function type `A2` above? Do you want to also have the static-side of `A` encoded into `A2`? What's an example of a use-case of the "non-abstract class declaration types" when dealing with abstract classes?

**Comment by tinganho** [comment]
> Do you want to also have the static-side of A encoded into A2

I think this is already handled?

> What's an example of a use-case of the "non-abstract class declaration types" when dealing with abstract classes

As I mentioned in the quoted comments above. I want pass a "non-abstract class declaration type" to a function and instantiate using new.

**Comment by aozgaa** [comment]
@tinganho 

Today, the following code will work:
```ts

        class A {}
        
        function factory(AA: typeof A) {
            return new AA;
        }
```

**Comment by tinganho** [comment]
@aozgaa thanks!
